File: packages/java/examples/ProfileMarks/ProfileComposition.java
package org.procedurals.examples.profilemarks;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.procedurals.mesh.RadialProfile3D;

/** Editable profiles for ProfileMarks; these settings are example choices, not library presets. */
public final class ProfileComposition {
    private final RadialProfile3D[] meshes;
    private final int slices;

    private ProfileComposition(RadialProfile3D[] meshes, int slices) {
        this.meshes = meshes;
        this.slices = slices;
    }

    public static ProfileComposition create(int slices, boolean capStart, boolean capEnd) {
        RadialProfile3D[] meshes = new RadialProfile3D[3];
        for (int shape = 0; shape < meshes.length; shape++) {
            List<Object> profile = new ArrayList<Object>();
            for (int point = 0; point <= 16; point++) {
                double fraction = point / 16.0;
                double z = -160.0 + point * 20.0;
                double radius = 60.0;
                if (shape == 1) radius -= 36.0 * Math.cos((fraction - 0.5) * Math.PI);
                if (shape == 2) radius *= 1.0 - fraction;
                // Replace these radius expressions or supply your own increasing [z,radius] pairs.
                profile.add(Arrays.<Object>asList(z, radius));
            }
            Map<String, Object> config = new LinkedHashMap<String, Object>();
            config.put("profile", profile);
            config.put("slices", slices);
            config.put("capStart", capStart);
            config.put("capEnd", capEnd);
            config.put("maxFaces", 10000);
            meshes[shape] = RadialProfile3D.generate(config);
        }
        return new ProfileComposition(meshes, slices);
    }

    public RadialProfile3D meshAt(int shape) { return meshes[shape]; }
    public int slices() { return slices; }
}

File: packages/python/examples/profile_marks/profile_composition.py
"""Editable retained ProfileMarks forms; settings are example choices, not presets."""
import math
from procedurals.radial_profile import RadialProfile3D

def create_profile_composition(slices, cap_start, cap_end):
    if type(slices) is not int or not 3 <= slices <= 715827881 or type(cap_start) is not bool or (type(cap_end) is not bool):
        raise ValueError('invalid example settings')
    meshes = []
    for shape in range(3):
        profile = []
        for point in range(17):
            fraction = point / 16.0
            radius = 60.0
            if shape == 1:
                radius -= 36.0 * math.cos((fraction - 0.5) * math.pi)
            if shape == 2:
                radius *= 1.0 - fraction
            profile.append([-160.0 + point * 20.0, radius])
        meshes.append(RadialProfile3D.generate({'profile': profile, 'slices': slices, 'capStart': cap_start, 'capEnd': cap_end, 'maxFaces': 10000}))
    return _Composition(tuple(meshes), slices)

class _Composition:
    __slots__ = ('_meshes', 'slices', 'size')

    def __init__(self, meshes, slices):
        self._meshes = meshes
        self.slices = slices
        self.size = 3

    def mesh_at(self, index):
        if type(index) is not int or not 0 <= index < 3:
            raise ValueError('invalid example shape')
        return self._meshes[index]
